core/src/consensus/fork_choice.rs

```diff
@@ -99,47 +114,196 @@ fn select_forks_failed_switch_threshold(
         // slots, it might be enough to get majority vote.
-        SwitchForkDecision::SameFork
-    } else {
-        // If we can't switch and our last vote was on a non-duplicate/confirmed slot, then
-        // reset to the the next votable bank on the same fork as our last vote,
-        // but don't vote.
+        return SwitchForkDecision::SameFork;
+    }
+
+    // If we can't switch and our last vote was on a non-duplicate/confirmed slot, then
+    // reset to the the next votable bank on the same fork as our last vote,
+    // but don't vote.
+
+    // We don't just reset to the heaviest fork when switch threshold fails because
+    // a situation like this can occur:
+
+    /* Figure 1:
+                slot 0
+                    |
+                slot 1
+                /        \
+    slot 2 (last vote)     |
+                |      slot 8 (10%)
+        slot 4 (9%)
+    */
+
+    // Imagine 90% of validators voted on slot 4, but only 9% landed. If everybody that fails
+    // the switch threshold abandons slot 4 to build on slot 8 (because it's *currently* heavier),
+    // then there will be no blocks to include the votes for slot 4, and the network halts
+    // because 90% of validators can't vote
+    info!(
+        "Waiting to switch vote to {},
+        resetting to slot {:?} for now,
+        switch proof stake: {},
+        threshold stake: {},
+        total stake: {}",
+        heaviest_bank_slot,
+        reset_bank.as_ref().map(|b| b.slot()),
+        switch_proof_stake,
+        total_stake as f64 * SWITCH_FORK_THRESHOLD,
+        total_stake
+    );
+    failure_reasons.push(HeaviestForkFailures::FailedSwitchThreshold(
+        heaviest_bank_slot,
+        switch_proof_stake,
+        total_stake,
+    ));
+    switch_fork_decision
+}
+
+fn select_candidate_vote_and_reset_banks<'a>(
+    heaviest_bank: &'a Arc<Bank>,
+    heaviest_bank_on_same_voted_fork: Option<&'a Arc<Bank>>,
+    progress: &'a ProgressMap,
+    tower: &'a Tower,
+    failure_reasons: &mut Vec<HeaviestForkFailures>,
+    initial_switch_fork_decision: SwitchForkDecision,
+) -> CandidateVoteAndResetBanks<'a> {
+    match initial_switch_fork_decision {
+        SwitchForkDecision::FailedSwitchThreshold(switch_proof_stake, total_stake) => {
+            // Even though we think we want to switch to a new, heavier fork, we
+            // actually want to continue voting on our current fork if our last
+            // vote was unable land.
+            let final_switch_fork_decision = select_forks_failed_switch_threshold(
+                heaviest_bank_on_same_voted_fork.map(|bank| bank.as_ref()),
+                progress,
+                tower,
+                heaviest_bank.slot(),
+                failure_reasons,
+                switch_proof_stake,
+                total_stake,
+                initial_switch_fork_decision,
+            );
+            let candidate_vote_bank = if final_switch_fork_decision.can_vote() {
```

Comment:
Whoops. Updated the old comments to the following:
```
// We need to "super" refresh our vote to the tip of our last
// voted fork because our last vote is unable to land. This is
// inferred by initially determining we can't vote but then
// determining we can vote on the same fork.
```
and:
```
// Just return the original vote candidate (the heaviest bank)
// for logging purposes. We can't actually vote on it, but this
// will allow us to check if there are any additional voting
// failures besides the switch threshold.
```